text/0000-api-version-compat.md

```diff
@@ -0,0 +1,122 @@
+# Meta
+[meta]: #meta
+- Name: Lifecycle API version support changes
+- Start Date: 2020-05-14
+- Author(s): [Emily Casey](https://github.com/ekcasey)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+
+# Summary
+[summary]: #summary
+
+To enable smoother migrations and upgrades this RFC proposes the following changes to the API version compatibility rules:
+
+1. When the lifecycle increments an API versions in the `0.x` version line, this shall interpreted as non-breaking
+1. The spec may make certain types of "breaking" changes in a `0.x` API version, so long as those changes do not preclude simultaneous lifecycle support for the previous minor or require collaborating buildpacks to implement the same API version.
+1. When the lifecycle implments a new `0.x` API minor version, it will treat the usage of features removed between minor spec versions as deprecated.
+
+# Motivation
+[motivation]: #motivation
+
+## Non-Breaking 0.x API Versions
+Originally we specified that all minor version increments in the `0.x` line would be interpreted as breaking changes, when applying [compatibility rules](https://github.com/buildpacks/spec/blob/master/buildpack.md#buildpacktoml-toml).
+This was the correct decision initially b/c we needed a way to indicate breaking changes before the APIs reached 1.0.
+However, this has become annoying over time as it adds considerable friction to the release of non-breaking features. All platforms and buildpacks must explicitly update to support the new APIs before pulling in a new lifecycle, even when a new minor would not break them.
+As the community of platform and buildpack authors grows and matures, this pain is magnified.
+
+Now that the APIs have stabilized we can realistically guarantee that we will not make breaking lifecycle changes before the implementation of the 1.0 APIs are released.
+
+## Deprecations
+We will allow the spec to remove features in a minor API and interpret this as a deprecated feature in the lifecycle.
+Previously, we never formalized rules around deprecations. This made sense because all API changes have been breaking.
+However, deprecations can aid platforms and buildpacks as they prepare to perform a major API upgrade.
+
+The lifecycle will support deprecation modes to allow platform/buildpack authors to curate the experience for users
+and/or filter for deprecations relevant to their use case.
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Spec
+The spec will use future `0.x` API release to indicate changes that can be implemented in a non-breaking fashion. These may include:
+- purely additive features
+- simple flag/file/key renames (the previous usage will be supported but deprecated by the lifecycle)
+- removal of features (these will be interpreted as deprecated by the lifecycle)
```

Comment:
Rather than remove features from the spec, I'd prefer that we mark them as deprecated. We could even move them to special section of the document (rather than keep them inline with non-deprecated features). I think this is important because it's the only way for buildpacks authors and platform operators to assuredly identify deprecated features (i.e. a warning might only appear when a particular code path is executed, and it's difficult to ensure that *every* possible code path is executed in testing). 

As prior art, consider the [Java API specification](https://docs.oracle.com/javase/8/docs/api/overview-summary.html). In particular, the deprecated methods on a class like [java.lang.Thread](https://docs.oracle.com/javase/8/docs/api/java/lang/Thread.html). There is a tab in the interface that lists all the deprecated methods.

Reply:
Adding deprecations to the spec was the path I originally went down but I was struggling to find much prior art (deprecations in specifications vs. implementations). The java API is a good example I didn't find. I am happy to make this change if we think the prior art is solid.